### ISSUE TITLE
🔥 Custom In App Messaging Shown Multiple Times

### ISSUE THREAD

**Issue by mjmitche** [invertase/react-native-firebase#5535]
## Issue

I recently set up an in-app messaging campaign in Firebase.

When I create a campaign whose Per-device frequency limit is set to Once per device for this campaign I see the In App Message every-time I launch the application.

Any help on this issue would greatly be appreciated.

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
# N/A
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [x] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
warn Package @react-native-firebase/in-app-messaging has been ignored because it contains invalid configuration. Reason: Cannot find module '@react-native-firebase/in-app-messaging/package.json'
Require stack:
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/config/resolveNodeModuleDir.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/releaseChecker/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/runServer.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/server.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/cliEntry.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/index.js
- /Users/michael/Documents/work/reactnative/node_modules/react-native/cli.js
- /usr/local/lib/node_modules/react-native-cli/index.js
warn Package @react-native-community/eslint-config has been ignored because it contains invalid configuration. Reason: Cannot find module '@react-native-community/eslint-config/package.json'
Require stack:
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/config/resolveNodeModuleDir.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/releaseChecker/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/runServer.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/server.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/cliEntry.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/index.js
- /Users/michael/Documents/work/reactnative/node_modules/react-native/cli.js
- /usr/local/lib/node_modules/react-native-cli/index.js
warn Package patch-package has been ignored because it contains invalid configuration. Reason: Cannot find module 'patch-package/package.json'
Require stack:
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/config/resolveNodeModuleDir.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/releaseChecker/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/runServer.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/server.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/cliEntry.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/index.js
- /Users/michael/Documents/work/reactnative/node_modules/react-native/cli.js
- /usr/local/lib/node_modules/react-native-cli/index.js
warn Package postinstall-postinstall has been ignored because it contains invalid configuration. Reason: Cannot find module 'postinstall-postinstall/package.json'
Require stack:
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/config/resolveNodeModuleDir.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/tools/releaseChecker/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/runServer.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/server/server.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/commands/index.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/cliEntry.js
- /Users/michael/Documents/work/reactnative/node_modules/@react-native-community/cli/build/index.js
- /Users/michael/Documents/work/reactnative/node_modules/react-native/cli.js
- /usr/local/lib/node_modules/react-native-cli/index.js
info Fetching system and libraries information...
(node:11178) Warning: Accessing non-existent property 'padLevels' of module exports inside circular dependency
(Use `node --trace-warnings ...` to show where the warning was created)
System:
    OS: macOS 11.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 378.47 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 15.10.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.5.3 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
    Android SDK:
      API Levels: 30
      Build Tools: 30.0.2, 30.0.3
      System Images: android-30 | Google APIs Intel x86 Atom
  IDEs:
    Android Studio: 4.2 AI-202.7660.26.42.7351085
    Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.13.1 => 16.9.0 
    react-native: 0.63.4 => 0.60.6 
  npmGlobalPackages:
    react-native-cli: 2.0.1
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `12.1.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - `N`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
I would try for a https://stackoverflow.com/help/minimal-reproducible-example here - my *guess* is that you are cycling the install token - I *believe* (assertion may be tested via an experiment) that the in app messaging is tied to the installation id (aka messaging.getToken() value etc) and if you cycle that you may see a "once per device" IAM multiple times?

I don't use IAM myself so don't have direct experience but that's my hunch. There are even cases of other libraries unexpectedly cycling the token for you, thus the minimal reproducible example to pin down an App.js that all by itself can cause this. If that shows an error here of course we can work to fix it

**Comment by mjmitche** [comment]
thank you so much @mikehardy! your comment is enlightening. and my apologies for not filling out my issue better.

I think your hunch may be exactly what's happening in my case. in App.js messaging().getToken() is getting called every time the app is re-ran. I'm going to add logic to only call that one time and am hoping that fixes the problem. thanks again!

**Comment by mikehardy** [comment]
Gettoken should be okay to run, I would look for any unexpected deletetoken calls, and/or compare results of token fetch to see if value changes (it should be stable). That'll be a good test of the hunch I think

**Comment by mikehardy** [comment]
Hey @mjmitche curious if you have any findings to share here? We're tracking a separate issue where messaging tokens and installation tokens appear unexpectedly related, and the question your issue raises - is the in-app-messaging token related to messaging token, and cycled on messaging.deleteToken? - is open. Curious for results

https://github.com/firebase/firebase-ios-sdk/issues/8491#issuecomment-892875261

**Comment by mikehardy** [comment]
I wonder if this is relevant, just released in firebase-ios-sdk 8.11.0: https://github.com/firebase/firebase-ios-sdk/pull/8922
